// astro/dist/core/messages.js
import stripAnsi from "strip-ansi";
import { bold, dim, red, green, underline, yellow, bgYellow, cyan, bgGreen, black } from "kleur/colors";
import { pad, emoji, getLocalAddress, getNetworkLogging } from "./dev/util.js";
import os from "os";
const PREFIX_PADDING = 6;
function req({ url, statusCode, reqTime }) {
  let color = dim;
  if (statusCode >= 500)
    color = red;
  else if (statusCode >= 400)
    color = yellow;
  else if (statusCode >= 300)
    color = dim;
  else if (statusCode >= 200)
    color = green;
  return `${bold(color(pad(`${statusCode}`, PREFIX_PADDING)))} ${pad(url, 40)} ${reqTime ? dim(Math.round(reqTime) + "ms") : ""}`.trim();
}
function reload({ file }) {
  return `${green(pad("reload", PREFIX_PADDING))} ${file}`;
}
function hmr({ file }) {
  return `${green(pad("update", PREFIX_PADDING))} ${file}`;
}
function devStart({
  startupTime,
  devServerAddressInfo,
  config,
  https,
  site
}) {
  const version = "0.24.3";
  const rootPath = site ? site.pathname : "/";
  const localPrefix = `${dim("\u2503")} Local    `;
  const networkPrefix = `${dim("\u2503")} Network  `;
  const { address: networkAddress, port } = devServerAddressInfo;
  const localAddress = getLocalAddress(networkAddress, config);
  const networkLogging = getNetworkLogging(config);
  const toDisplayUrl = (hostname) => `${https ? "https" : "http"}://${hostname}:${port}${rootPath}`;
  let addresses = [];
  if (networkLogging === "none") {
    addresses = [`${localPrefix}${bold(cyan(toDisplayUrl(localAddress)))}`];
  } else if (networkLogging === "host-to-expose") {
    addresses = [`${localPrefix}${bold(cyan(toDisplayUrl(localAddress)))}`, `${networkPrefix}${dim("use --host to expose")}`];
  } else {
    addresses = Object.values(os.networkInterfaces()).flatMap((networkInterface) => networkInterface ?? []).filter((networkInterface) => (networkInterface == null ? void 0 : networkInterface.address) && (networkInterface == null ? void 0 : networkInterface.family) === "IPv4").map(({ address }) => {
      if (address.includes("127.0.0.1")) {
        const displayAddress = address.replace("127.0.0.1", localAddress);
        return `${localPrefix}${bold(cyan(toDisplayUrl(displayAddress)))}`;
      } else {
        return `${networkPrefix}${bold(cyan(toDisplayUrl(address)))}`;
      }
    }).sort((msg) => msg.startsWith(localPrefix) ? -1 : 1);
  }
  const messages = [`${emoji("\u{1F680} ", "")}${bgGreen(black(` astro `))} ${green(`v${version}`)} ${dim(`started in ${Math.round(startupTime)}ms`)}`, "", ...addresses, ""];
  return messages.map((msg) => `  ${msg}`).join("\n");
}
function prerelease({ currentVersion }) {
  const tag = currentVersion.split("-").slice(1).join("-").replace(/\..*$/, "");
  const badge = bgYellow(black(` ${tag} `));
  const headline = yellow(`\u25B6 This is a ${badge} prerelease build`);
  const warning = `  Feedback? ${underline("https://astro.build/issues")}`;
  return [headline, warning, ""].map((msg) => `  ${msg}`).join("\n");
}
function portInUse({ port }) {
  return `Port ${port} in use. Trying a new one\u2026`;
}
function err(error) {
  if (!error.stack)
    return stripAnsi(error.message);
  let message = stripAnsi(error.message);
  let stack = stripAnsi(error.stack);
  const split = stack.indexOf(message) + message.length;
  message = stack.slice(0, split);
  stack = stack.slice(split).replace(/^\n+/, "");
  return `${message}
${dim(stack)}`;
}
export {
  devStart,
  err,
  hmr,
  portInUse,
  prerelease,
  reload,
  req
};
